src/vmm/src/arch/aarch64/regs.rs

```diff
@@ -157,265 +110,551 @@ arm64_sys_reg!(ID_AA64MMFR2_EL1, 3, 0, 0, 7, 2);
+            vec: self,
+        }
+    }
+}
+
+/// Iterator over `Aarch64RegisterVec`.
+#[derive(Debug)]
+pub struct Aarch64RegisterVecIterator<'a> {
+    index: usize,
+    offset: usize,
+    vec: &'a Aarch64RegisterVec,
 }
 
-/// Read the MPIDR - Multiprocessor Affinity Register.
-pub fn get_mpidr(vcpufd: &VcpuFd) -> Result<u64, Error> {
-    match vcpufd.get_one_reg(MPIDR_EL1) {
-        Err(err) => Err(Error::GetOneReg(MPIDR_EL1, err)),
-        // MPIDR register is 64 bit wide on aarch64, this expect cannot fail
-        // on supported architectures
-        Ok(val) => Ok(val.try_into().expect("MPIDR register to be 64 bit")),
+impl<'a> Iterator for Aarch64RegisterVecIterator<'a> {
+    type Item = Aarch64RegisterRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.vec.ids.len() {
+            let id = self.vec.ids[self.index];
+            let reg_size = reg_size(id) as usize;
+            let reg_ref = Aarch64RegisterRef {
+                id,
+                data: &self.vec.data[self.offset..self.offset + reg_size],
+            };
+            self.index += 1;
+            self.offset += reg_size;
+            Some(reg_ref)
+        } else {
+            None
+        }
     }
 }
 
-/// Saves the states of the system registers into `state`.
-///
-/// # Arguments
-///
-/// * `regs` - Input/Output vector of registers.
-pub fn get_all_registers(vcpufd: &VcpuFd, state: &mut Vec<Aarch64Register>) -> Result<(), Error> {
-    get_registers(vcpufd, &get_all_registers_ids(vcpufd)?, state)
+/// Iterator over `Aarch64RegisterVec` with mutable values.
+#[derive(Debug)]
+pub struct Aarch64RegisterVecIteratorMut<'a> {
+    index: usize,
+    offset: usize,
+    vec: &'a mut Aarch64RegisterVec,
 }
 
-/// Saves states of registers into `state`.
-///
-/// # Arguments
-///
-/// * `ids` - Slice of registers ids to save.
-/// * `regs` - Input/Output vector of registers.
-pub fn get_registers(
-    vcpufd: &VcpuFd,
-    ids: &[u64],
-    regs: &mut Vec<Aarch64Register>,
-) -> Result<(), Error> {
-    for id in ids.iter() {
-        regs.push(Aarch64Register {
-            id: *id,
-            value: vcpufd
-                .get_one_reg(*id)
-                .map_err(|e| Error::GetOneReg(*id, e))?,
-        });
+impl<'a> Iterator for Aarch64RegisterVecIteratorMut<'a> {
+    type Item = Aarch64RegisterRefMut<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.vec.ids.len() {
+            let id = self.vec.ids[self.index];
+            let reg_size = reg_size(id) as usize;
+            // # Safety:
+            // Registers data packed in the `vec.data`
+            // The offset for each register data is a
+            // sum of sizes of previous registers sizes.
+            let data = unsafe {
+                let data = self.vec.data.as_mut_ptr();
+                let reg_data = data.add(self.offset);
+                std::slice::from_raw_parts_mut(reg_data, reg_size)
+            };
```

Comment:
Looking at this more closely, since this is deserialized data from a snapshot, and there is no (from what I can tell) validation in place that ensures that `vec.data` and `vec.ids` are in sync (e.g. someone might modify the snapshot file such that there are more reg ids than data in the packed register store), this can lead to an out of bounds memory access. I think we need a check somewhere (during deserialization, I suppose) that ensures that the packed data store has the correct length for the list of IDs given. 

Particularly, please extend the `SAFETY` comment to address all invariants of [`std::ptr::add`](https://doc.rust-lang.org/std/primitive.pointer.html#safety-7) and [`std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety)

Reply:
Updated `Safety` comments.
To add verification during deserialization I added a wrapper type around `Aarch64RegisterVec` (now known as `Aarch64RegisterVecInner`).